src/wp-includes/functions.php

```diff
@@ -2519,40 +2519,55 @@ function wp_unique_filename( $dir, $filename, $unique_filename_callback = null )
 			$filename = str_replace( "{$fname}{$ext}", "{$fname}-{$number}{$ext}", $filename );
 		}
 
-		// Change '.ext' to lower case.
-		if ( $ext && strtolower( $ext ) != $ext ) {
-			$ext2      = strtolower( $ext );
-			$filename2 = preg_replace( '|' . preg_quote( $ext ) . '$|', $ext2, $filename );
-
-			// Check for both lower and upper case extension or image sub-sizes may be overwritten.
-			while ( file_exists( $dir . "/{$filename}" ) || file_exists( $dir . "/{$filename2}" ) ) {
-				$new_number = (int) $number + 1;
-				$filename   = str_replace( array( "-{$number}{$ext}", "{$number}{$ext}" ), "-{$new_number}{$ext}", $filename );
-				$filename2  = str_replace( array( "-{$number}{$ext2}", "{$number}{$ext2}" ), "-{$new_number}{$ext2}", $filename2 );
-				$number     = $new_number;
-			}
+		// Get the mime type. Uploaded files were already checked with wp_check_filetype_and_ext()
+		// in _wp_handle_upload(). Using wp_check_filetype() would be sufficient here.
+		$file_type = wp_check_filetype( $filename );
+		$mime_type = $file_type['type'];
 
-			$filename = $filename2;
-		} else {
-			while ( file_exists( $dir . "/{$filename}" ) ) {
-				$new_number = (int) $number + 1;
+		$is_image    = ( ! empty( $mime_type ) && 0 === strpos( $mime_type, 'image/' ) );
+		$upload_dir  = wp_get_upload_dir();
+		$lc_filename = null;
 
-				if ( '' === "{$number}{$ext}" ) {
-					$filename = "{$filename}-{$new_number}";
-				} else {
-					$filename = str_replace( array( "-{$number}{$ext}", "{$number}{$ext}" ), "-{$new_number}{$ext}", $filename );
-				}
+		$lc_ext = strtolower( $ext );
+		$_dir   = trailingslashit( $dir );
+
+		// If the extension is uppercase add an alternate file name with lowercase extension. Both need to be tested
+		// for uniqueness as the extension will be changed to lowercase for better compatibility with different filesystems.
+		if ( $ext && $lc_ext !== $ext ) {
```

Comment:
> have you considered just setting the extension as lowercase early in the function

Unfortunately `$ext` is passed as arg in the `wp_unique_filename` filter at the bottom. Changing it might affect something and would be considered a regression.

> files with uppercase extensions exist that were added by some other means

That's not impossible but never seen a bug report about it, and is generally not supported. Files added "by hand" to the upload dirs would not have attachment posts and will be unusable from whitin WP. Perhaps if there's a bug report/ticket it can be fixed there.

Reply:
> Unfortunately `$ext` is passed as arg in the `wp_unique_filename` filter at the bottom. Changing it might affect something and would be considered a regression.

Ah, good shout, I'll fix my patch for that.

> That's not impossible but never seen a bug report about it, and is generally not supported.

We've seen this very scenario a bunch of times in support, either via a dump of media from Windows and then sucked in to a gallery plugin that doesn't use the Media Library, or imported into the Media Library from the filesystem by some importer.

Reply:
> We've seen this very scenario a bunch of times

Did it result in lost/overwritten files? Sure, lets fix it but in another ticket and better not in a dot release.

Hm, actually, wouldn't `file_exists()` catch possible collisions caused by upper/lower case? It does a filesystem call so thinking it will.

Reply:
> Hm, actually, wouldn't `file_exists()` catch possible collisions caused by upper/lower case? It does a filesystem call so thinking it will.

No, on a case sensitive filesystem (effectively all Linux servers) `file_exists( 'wibble.png' )` will not see a `wibble.PNG` file.